ui-framework-documentation/layers.md

```diff
@@ -0,0 +1,214 @@
+    ...
+    }
+    ```
+* To make an action depending on a value:
+
+  
+    ```perl
+    sub create_filesystem {
+        my ($self, $filesystem) = @_;
+        $self->get_filesystem_options_page()->select_filesystem($filesystem);
+    ...
+    }
+    ```
+    
+### Access to os-autoinst testapi from Controller
+
+Ideally, it should not use the os-autoinst testapi directly. 
+
+* Do not use testapi methods that communicates with the system under
+  test (e.g. `send_keys`, `assert_screen`). Wrap them into [Page](#page)
+  methods with the meaningful names instead.
+
+* Using `get_var` to change the flow of a test or get a data for the
+  test should be avoided as much as possible (e.g. to decide whether
+  check or uncheck checkbox, use method parameters instead and pass the
+  data from [Test Module](#test-module).
+
+**NOTE:** Since os-autoinst API does not have separation between methods
+for interacting with openQA and interacting with the system under test,
+the usage of os-autoinst testapi in [Controller](#controller) layer
+cannot be fully avoided at the moment. So, some testapi functions may be
+used in exceptional cases.
+
+#### Example:
+
+It is ok when `get_var` is used in combination with
+`record_soft_failure` to just highlight known issue in openQA.
+
+```perl
+sub create_encrypted_partition {
+...
+    if (get_var('SOME_VARIABLE')) {
+        record_soft_failure('bsc#1234567');
+        return;
+    }
+...
+}
+```
+
+### Access to other framework layers from Controller
+
+It knows only about [Pages](#page) and hides the complexity of
+manipulating with them from the [Test Module](#test-module) layer.
+
+However, it also provides access to the [Pages](#page) directly through
+the getters. This compromising solution was added for the cases when
+specific and rare actions should be made.
+
+#### Example:
+  
+For instance, there might be a test, that should create an encrypted
+partition. Positive case may use something like:
+
+```perl
+sub create_encrypted_partition {
+    my ($self) = @_;
+    $self->get_partitioning_scheme_page()->select_enable_disk_encryption_checkbox();
+    $self->get_partitioning_scheme_page()->enter_password();
+    $self->get_partitioning_scheme_page()->enter_password_confirmation();
+    $self->get_partitioning_scheme_page()->press_next();
+}
+```
+
+Then it could be called in all the [Test Modules](#test-module), where 
+the encrypted partition need to be created.
+
+```perl
+sub run {
+    my $partitioner = $testapi::distri->get_partitioner();
+    $partitioner->create_encrypted_partition();
+}    
+``` 
+
+But in case if need to verify some negative cases the method cannot be
+used as is. For instance, to verify that prompt appears when blank
+password is entered. For such kind of cases the access to the
+[Pages](#page) through [Controller](#controller) is added.  
+Then [Test Module](#test-module) code may look like:
+
+```perl
+sub run {
+    my $partitioner = $testapi::distri->get_partitioner();
```

Comment:
perhaps to put all together we might want to name controlers? for example get_partitioner_controller? Page concept seems more clear to me than Controller in the examples. make sense?

Reply:
Yes, makes sense.  Also, for the particular case, it probably should even be something like 'edit_suggested_partitioning_controller', as we also have Expert Partitioner.

Naming part is still under consideration. For example, I'm not sure what is better name for controller layer - "Controller" or, for example "Steps".